### PR TITLE
Fix force speculation

### PR DESCRIPTION
**Pull request by JanJecmen**
Attempt at fixing #577

### CODE REVIEW

**Comment by JanJecmen** [comment]
@o- This doesn't propagate the feedback to the argument of force, it instead marks the force with a flag and during type speculation temporarily adds the force's type feedback to its argument's. This should avoid the issue completely, if I understand it correctly?

As a side note, I remember you saying that the checkpoint selection in type speculation was wrong (shouldn't use `next()`) but I cannot recall what was the conclusion (and why it ended up working when that part wasn't changed)

File: rir/src/compiler/opt/type_speculation.cpp [hunk 1]
```diff
@@ -45,13 +45,18 @@ void TypeSpeculation::apply(RirCompiler&, ClosureVersion* function,
...
-                        i = Instruction::Cast(i->arg(0).val());
-                    break;
-                }
-                default: {}
+                    if (force->observed != Force::ArgumentKind::unknown &&
+                        force->observed != Force::ArgumentKind::promise &&
+                        Instruction::Cast(force->input())) {
+                        i = Instruction::Cast(force->input());
+                        if (force->observed == Force::ArgumentKind::value)
+                            i->typeFeedback = force->typeFeedback;
```
Comment (o-) [inline rir/src/compiler/opt/type_speculation.cpp]: drive by comment: wouldn't this do the exact same thing as the previous try, just a bit later? if multiple force have the same input then this still leads to wrong typefeeback.

**Comment by JanJecmen** [comment]
Update: the `while vector` benchmark times out now, probably because the vector is copied in every iteration (we have `subassign` - named = 0, then `ensure named` - named = 1, then `store` - named = 2)

Didn't get further in investigating this

**Comment by o-** [comment]
that means that the result of subassign is consumed twice in pir.

do you see speedups?

**Comment by o-** [comment]
#607 might help. actually wanted to do that already a while ago

**Comment by JanJecmen** [comment]
Looks like that could help
Didn't check performance yet, just worked to get it fixed

**Pull request merged**